Adding support for `fp16` for asr pipeline.

Narsil opened the pull request:
# What does this PR do?

Fixes #20862 

Many things were considered before settling for this design.

- `feature_extractor(return_tensors="pt¨, torch_dtype=torch_dtype)` . This would have the advantage of being consistent, but not all feature extractors to define this, so it would affect all of them. Then why would we use `torch_dtype` instead of the more common place `dtype` which could be applied to TF and flax as well. Also it feels a bit redundant to specify both` return_tensors` and `torch_dtype`, it would be a good candidate to fuse both parameters (but outisde the scope of this PR).
- `AutoFeatureExtractor.from_pretrained(..., torch_dtype=torch_dtype)`. This would have the advantage of being overall so users don't need to respecify on each call. However we can't specifiy `return_tensors="pt" ` in there either, so for consistency I didn't try to put it there.
- `ffmpeg_read(..., dtype=dtype)` This would be nice to load directly the waveform into fp16 and just let fp16 flow through the feature_extractor. However, whisper in particular uses mel_spectrogram, so using f16 sound might actually damage performance.

In the end, this solution is the simplement I could come up with. Let `torch_dtype` flow to the pipeline, use it as a regular parameter and convert the output of the feature_extractor after.

This does incur a potentially extra copy but there's no risk of damaging quality of the input.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts and @NielsRogge
- speech models: @sanchit-gandhi

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger and @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

Narsil committed:
Supporting `fp16` for asr pipeline

Narsil committed:
Adding test.

Narsil committed:
Style.

Narsil committed:
Oops.

Narsil committed:
Flake8 update ?

Narsil committed:
Fixing flake8 ?

Narsil committed:
Revert "Flake8 update ?"

This reverts commit 0b917fcb520e5f34d1933d9d37d8f32b64553048.

Narsil committed:
Style (acctidentally deleted flake8 F401.)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20864). All of your documentation changes will be reflected on that endpoint.

Narsil committed:
Move to a bigger test (no small whisper model, and s2t doesn't seem to
accept torch_dtype=fp16).

Also we need to use a GPU to actually compute on fp16.

bofenghuang commented on src/transformers/pipelines/automatic_speech_recognition.py:
```diff
@@ -332,12 +337,14 @@ def preprocess(self, inputs, chunk_length_s=0, stride_length_s=None, ignore_warn
                 inputs, sampling_rate=self.feature_extractor.sampling_rate, return_tensors="pt"
             )
+            if dtype is not None:
+                processed = {k: v.to(dtype=dtype) for k, v in processed.items()}
```
Hi @Narsil,

This works fine for `seq2seq` models. 

But in case of `ctc` or `ctc_with_lm` model type, I think we should just cast `input_values` from `float32` to `float16`, and leave `attention_mask` to `int32` if it exists.

sgugger replied:
Yes. And as above, if you directly use the `to` method on `processed`, it will take care of that for you.

Narsil replied:
Done. Thanks, TIL

sgugger commented on src/transformers/pipelines/automatic_speech_recognition.py:
```diff
@@ -52,13 +52,15 @@ def rescale_stride(stride, ratio):
         chunk = inputs[i : i + chunk_len]
         processed = feature_extractor(chunk, sampling_rate=feature_extractor.sampling_rate, return_tensors="pt")
+        if dtype is not None:
+            processed = {k: v.to(dtype=dtype) for k, v in processed.items()}
```
I believe you can call the `to` directly on `processed`, which is a `BatchFeature` and handles `dtype` in its `to` method thanks to #20536 (was designed for vision but I think it will apply here too).

Narsil replied:
Done !

sgugger commented on src/transformers/pipelines/automatic_speech_recognition.py:
```diff
@@ -249,7 +253,8 @@ def _sanitize_parameters(self, **kwargs):
 
-    def preprocess(self, inputs, chunk_length_s=0, stride_length_s=None, ignore_warning=False):
+    def preprocess(self, inputs, chunk_length_s=0, stride_length_s=None, ignore_warning=False, dtype=None):
+        print(f"Running with dtype {dtype}")
```
To be cleaned up ;-)

Narsil replied:
Oops

sgugger reviewed (approved):
Thanks for working on this! My only comment is to make sure to leverage the `to` method on `BatchFeature` (if the feature extractor here returns another type, maybe make sure its `to` method handles dtype arguments) so that checks like not converting int inputs are applied for free.

Otherwise LGTM!

Narsil committed:
Using BatchFeature capability.
